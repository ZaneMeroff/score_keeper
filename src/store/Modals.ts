// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Modals extends VuexModule {

  addPlayersModal: boolean = false
  clearDataModal: boolean = false
  deletePlayersModal: boolean = false
  resetScoreModal: boolean = false
  scoreLimitsModal: boolean = false

  get getShowAddPlayersModal(): boolean {
    return this.addPlayersModal
  }

  get getShowClearDataModal(): boolean {
    return this.clearDataModal
  }

  get getShowDeletePlayersModal(): boolean {
    return this.deletePlayersModal
  }

  get getShowResetScoreModal(): boolean {
    return this.resetScoreModal
  }

  get getShowScoreLimitsModal(): boolean {
    return this.scoreLimitsModal
  }

  // ------------------------------------------
  //       add players modal visibility
  // ------------------------------------------
  
  @Mutation
  addPlayersModalVisibility(show: boolean) {
    this.addPlayersModal = show
  }
  
  @Action
  action_addPlayersModalVisibility(show: boolean) {
    this.addPlayersModalVisibility(show)
  }

  // ------------------------------------------
  //        clear data modal visibility
  // ------------------------------------------
  
  @Mutation
  clearDataModalVisibility(show: boolean) {
    this.clearDataModal = show
  }
  
  @Action
  action_clearDataModalVisibility(show: boolean) {
    this.clearDataModalVisibility(show)
  }

  // ------------------------------------------
  //  confirm delete players modal visibility
  // ------------------------------------------
  
  @Mutation
  deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModal = show
  }

  @Action
  action_deletePlayersModalVisibility(show: boolean) {
    this.deletePlayersModalVisibility(show)
  }

  // ------------------------------------------
  //    confirm reset score modal visibility
  // ------------------------------------------
  
  @Mutation
  resetScoreModalVisibility(show: boolean) {
    this.resetScoreModal = show
  }

  @Action
  action_resetScoreModalVisibility(show: boolean) {
    this.resetScoreModalVisibility(show)
  }

  // ------------------------------------------
  //        score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitsModal = show
  }

  @Action
  action_scoreLimitModalVisibility(show: boolean) {
    this.scoreLimitModalVisibility(show)
  }
}