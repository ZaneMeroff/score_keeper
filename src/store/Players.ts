// Vue & Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Player, PlayerData } from '@/types/players'

@Module
export default class Players extends VuexModule {

    playerData: PlayerData = {}

    get getPlayerData() {
      return this.playerData
    }

    // ------------------------------------------
    //              create player
    // ------------------------------------------

    @Mutation
    createPlayer(payload: Player) {
      Vue.set(this.playerData, payload.id, payload)
    }

    @Action
    action_createPlayer(payload: Player) {
      this.createPlayer(payload)
    }

    // ------------------------------------------
    //         modify target player score
    // ------------------------------------------

    @Mutation
    setPlayerScore(payload: { id: string, score: number }) {
      this.playerData[payload.id].score = payload.score
    }

    @Action
    action_setPlayerScore(payload: { id: string, score: number }) {
      this.setPlayerScore(payload)
    }

    // ------------------------------------------
    //         modify target player name
    // ------------------------------------------

    @Mutation
    setPlayerName(payload: { id: string, name: string }) {
      this.playerData[payload.id].name = payload.name
    }

    @Action
    action_setPlayerName(payload: { id: string, name: string }) {
      this.setPlayerName(payload)
    }

    // ------------------------------------------
    //        zero scores for all players
    // ------------------------------------------

    @Mutation
    zeroScores() {
      for (const player in this.playerData) {
        this.playerData[player].score = 0
      }
    }

    @Action
    action_zeroScores() {
      this.zeroScores()
    }
}