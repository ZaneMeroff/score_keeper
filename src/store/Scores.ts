// Decorators
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

// Types
import { Score, ScoreData } from '@/types/scores'
import Vue from 'vue'

@Module
export default class Scores extends VuexModule {

    scoreData: ScoreData = {}

    get getScoreData() {
      return this.scoreData
    }

    @Mutation
    setScoreData(payload: { id: string, updatedScore: number }) {
      // create or update score data based on id
      Vue.set(this.scoreData, payload.id, payload.updatedScore)
    }

    
}