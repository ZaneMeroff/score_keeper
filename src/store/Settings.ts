// Decorators
import Vue from 'vue'
import { Action, Module, Mutation, VuexModule } from 'vuex-class-modules'

@Module
export default class Settings extends VuexModule {

  darkMode: boolean = true
  minScore: number = 0
  maxScore: number = 10
  showScoreLimitsModal: boolean = false
  // showMenuDrawer: boolean = false

  get getDarkMode() {
    return this.darkMode
  }

  get getMinScore() {
    return this.minScore
  }

  get getMaxScore() {
    return this.maxScore
  }

  get getShowScoreLimitsModal() {
    return this.showScoreLimitsModal
  }

  // get getShowMenuDrawer() {
  //   return this.showMenuDrawer
  // }
  
  // ------------------------------------------
  //             toggle dark mode
  // ------------------------------------------

  @Mutation
  setDarkMode() {
    this.darkMode = !this.darkMode
  }

  @Action
  action_setDarkMode() {
    this.setDarkMode()
  }

  // ------------------------------------------
  //            set score limits
  // ------------------------------------------

  @Mutation
  setScoreLimits(payload: { min: number, max: number }) {
    this.minScore = payload.min
    this.maxScore = payload.max
  }

  @Action
  action_setScoreLimits(payload: { min: number, max: number }) {
    this.setScoreLimits(payload)
  }

  // ------------------------------------------
  //       score limit modal visibility
  // ------------------------------------------
  
  @Mutation
  setScoreLimitModalVisibility(show: boolean) {
    this.showScoreLimitsModal = show
  }

  @Action
  action_setScoreLimitModalVisibility(show: boolean) {
    this.setScoreLimitModalVisibility(show)
  }

  // ------------------------------------------
  //         menu drawer visibility
  // ------------------------------------------
  
  // @Mutation
  // setMenuDrawerlVisibility(show: boolean) {
  //   this.showMenuDrawer = show
  // }

  // @Action
  // action_setMenuDrawerlVisibility(show: boolean) {
  //   this.setMenuDrawerlVisibility(show)
  // }
}