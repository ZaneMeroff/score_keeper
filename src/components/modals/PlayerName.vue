<template>

 <q-dialog v-model="showModal" persistent>
    <q-card class="q-card" :dark="stateSettings.getDarkMode">

      <q-card-section>
        <div class="q-mb-md">

          <!-- header -->
          <div class="text q-mb-sm" :style="stateSettings.getDarkModeText">Edit Name</div>
          <q-separator color="blue" class="q-mb-lg" inset></q-separator>
          
          <!-- text input -->
          <q-input 
            v-model="name" 
            maxlength="20"
            style="font-size: 16px"
            :dark="stateSettings.getDarkMode"
            :rules="[ val => !!val || '* Required' ]"
            @input="onInputChange"
          />
        </div>

        <q-btn-group flat class="q-btn-group">
          <!-- save button -->
          <q-btn
            flat
            class="q-btn"
            color="green"
            id="save-btn"
            size="md"
            :disable="disabled"
            :label="'save'"
            @click="handleSaveBtn"
          />
          <!-- cancel button -->
          <q-btn
            flat
            class="q-btn"
            color="red"
            size="md"
            :label="'cancel'"
            @click="handleCancelBtn"
          />
        </q-btn-group>
      </q-card-section>
    </q-card>
  </q-dialog>

</template>

<script lang="ts">

// Decorators
import { Component, Prop, Vue } from 'vue-property-decorator';

// Vuex
import { statePlayers, stateSettings } from '@/store/index'
import Players from '@/store/Players'
import Settings from '@/store/Settings'

@Component
export default class PlayerName extends Vue {
  
  @Prop(Boolean) private showModal!: boolean
  @Prop(String) private playerId!: string

  statePlayers: Players = statePlayers
  stateSettings: Settings = stateSettings

  name: string = ''
  disabled: boolean = false

  handleSaveBtn() {
    this.statePlayers.action_setPlayerName({ id: this.playerId, name: this.name })
    this.setDefault()
    this.$emit('close')
  }

  handleCancelBtn() {
    this.setDefault()
    this.$emit('close')
  }

  onInputChange() {
    if (this.name.length) this.disabled = false
    else this.disabled = true
  }

  setDefault() {
    if (this.statePlayers.getPlayerData[this.playerId]) {
      this.name = this.statePlayers.getPlayerData[this.playerId].name
    }
  }

  mounted() {
    this.setDefault()
  }
}
</script>

<style scoped>

  .q-card {
    width: 300px;
  }

  .q-btn-group {
    display: inline !important;
  }

  .q-btn {
    width: 50%;
  }

  .text {
    font-size: 16px;
    text-align: center;
  }

</style>