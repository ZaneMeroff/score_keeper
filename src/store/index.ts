import Vue from 'vue'
import Vuex from 'vuex'
import VuexPersistence from 'vuex-persist'
import localforage from 'localforage'

import Modals from './Modals'
import Players from './Players'
import Settings from './Settings'

Vue.use(Vuex)

const vuexLocal = new VuexPersistence({
  storage: localforage as any,
  asyncStorage: true,
  modules: [ 'Modals', 'Players', 'Settings' ],
});

export const store: any = new Vuex.Store({
  // strict: process.env.DEV === 'true',
  state: {},
  modules: {},
  plugins: [ vuexLocal.plugin ],
})

export const stateModals= new Modals({ store, name: 'Modals' })
export const statePlayers= new Players({ store, name: 'Players' })
export const stateSettings = new Settings({ store, name: 'Settings' })
