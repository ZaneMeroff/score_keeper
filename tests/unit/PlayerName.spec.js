import { createLocalVue, shallowMount } from '@vue/test-utils'
import { statePlayers } from '@/store/index'
import PlayerName from '@/components/modals/PlayerName.vue'
import Quasar, * as All from 'quasar'
import uuid from 'uuid/v4'
import Vuex from 'vuex'

// --------- Vuex & Quasar setup -------------

const localVue = createLocalVue()
localVue.use(Vuex, Quasar, {components: All, directives: All, plugins: All})

// --------------- Mocks ---------------------

jest.mock('uuid/v4')
uuid.mockImplementation(() => '12345')

// --------------- Setup ---------------------

statePlayers.action_createPlayers(1)

// -------------------------------------------

describe('PlayerName', () => {

  describe('snapshots', () => {

    it('should render with default html', () => {
      const propsData = { playerId: '12345', showModal: true }
      const component = shallowMount(PlayerName, { localVue, propsData })
      
      expect(component.element).toMatchSnapshot()
    })
  })

  describe('props', () => {

    describe('showModal', () => {

      it('should accept a boolean', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$props.showModal).toEqual(propsData.showModal)
      })
    })

    describe('playerId', () => {

      it('should accept a string', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$props.playerId).toEqual(propsData.playerId)
      })
    })
  })

  describe('data', () => {

    describe('name', () => {

      it('should have a default value of empty string', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        const expected = 'Player 1'
        expect(component.vm.$data.name).toEqual(expected)
      })

      it('should accept a string', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = 'New Name'

        component.setData({ name: newValue })
        expect(component.vm.$data.name).toEqual(newValue)
      })
    })

    describe('disabled', () => {

      it('should have a default value of false', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        expect(component.vm.$data.disabled).toEqual(false)
      })

      it('should accept a boolean', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = true

        component.setData({ disabled: newValue })

        expect(component.vm.$data.disabled).toEqual(newValue)
      })
    })
  })

  describe('methods', () => {

    describe('handleSaveBtn', () => {
      
      it.skip('should call statePlayers.action_setPlayerName', () => {
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
      })

      it('should call setDefault', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const spy = jest.spyOn(component.vm, 'setDefault')

        component.vm.handleSaveBtn()
        expect(spy).toHaveBeenCalledTimes(1)
      })

      it('should emit close event', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        component.vm.handleSaveBtn()
        const expected = { 'close': [[]] }
        expect(component.emitted()).toEqual(expected)
      })
    })

    describe('handleCancelBtn', () => {
      
      it('should call setDefault', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const spy = jest.spyOn(component.vm, 'setDefault')

        component.vm.handleCancelBtn()
        expect(spy).toHaveBeenCalledTimes(1)
      })

      it('should emit close event', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        component.vm.handleCancelBtn()
        const expected = { 'close': [[]] }
        expect(component.emitted()).toEqual(expected)
      })
    })

    describe('onInputChange', () => {
      
      it('should set disabled to true if name.length=false', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })

        component.setData({ name: '' })
        component.vm.onInputChange()
        expect(component.vm.$data.disabled).toEqual(true)
      })

      it('should set disabled to false if name.length=true', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const newValue = 'Sample Name'

        component.setData({ name: newValue })
        component.vm.onInputChange()
        expect(component.vm.$data.disabled).toEqual(false)
      })
    })

    describe('setDefault', () => {
      
      it('should set name correctly', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        
        component.vm.setDefault()
        const expected = 'Player 1'
        expect(component.vm.$data.name).toEqual(expected)
      })
    })
  })

  describe('lifecycle methods', () => {

    describe('mounted', () => {

      it.skip('should call setDefault on mount', () => {
        const propsData = { playerId: '12345', showModal: true }
        const component = shallowMount(PlayerName, { localVue, propsData })
        const spy = jest.spyOn(component.vm, 'setDefault')
        // ************************************************
        // ************************************************
        // ***************  incomplete  *******************
        // ************************************************
        // ************************************************
        expect(spy).toHaveBeenCalledTimes(1)
      })
    })
  })
})
