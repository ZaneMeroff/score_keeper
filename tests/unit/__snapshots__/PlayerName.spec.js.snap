// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerName snapshots should render with default html 1`] = `
<q-dialog-stub
  persistent="true"
  position="standard"
  value="true"
>
  <q-card-stub
    class="q-card"
    dark="true"
    tag="div"
  >
    <q-card-section-stub
      tag="div"
    >
      <div
        class="q-mb-md"
      >
        <div
          class="text q-mb-sm"
          style="color: rgb(255, 255, 255);"
        >
          Edit Name
        </div>
         
        <q-separator-stub
          class="q-mb-lg"
          color="blue"
          inset="true"
        />
         
        <q-input-stub
          dark="true"
          maxlength="20"
          rules="[Function]"
          style="font-size: 16px;"
          type="text"
          value=""
        />
      </div>
       
      <q-btn-group-stub
        class="q-btn-group"
        flat="true"
      >
        <q-btn-stub
          align="center"
          class="q-btn"
          color="green"
          flat="true"
          label="save"
          ripple="true"
          size="md"
        />
         
        <q-btn-stub
          align="center"
          class="q-btn"
          color="red"
          flat="true"
          label="cancel"
          ripple="true"
          size="md"
        />
      </q-btn-group-stub>
    </q-card-section-stub>
  </q-card-stub>
</q-dialog-stub>
`;
